src/lib/reasoners/rel_utils.ml

```diff
@@ -197,182 +197,210 @@ module type Constraint = sig
+      ) c acc
+
+  let add ?(pending = false) ~ex c t =
+    let c = explained ~ex c in
+    (* Note: use [CS.find] here, not [CS.mem], to ensure we use the same
+       explanation for [c] in the [pending] and [active] sets. *)
+    match CS.find c t.active with
+    | c ->
+      if pending then { t with pending = CS.add c t.pending } else t
+    | exception Not_found ->
+      let active = CS.add c t.active in
+      let args_map =
+        Constraint.fold_args (cs_add c) c.value t.args_map
+      in
+      let leaves_map = fold_leaves (cs_add c) c.value t.leaves_map in
+      let pending = CS.add c t.pending in
+      { active; args_map; leaves_map; pending }
+
+  let cs_remove c r cs_map =
     MX.update r (function
-        | Some css ->
-          let css = CS.remove cs css in
-          if CS.is_empty css then None else Some css
-        | None ->
-          (* Can happen if the same argument is repeated *)
-          None
+        | Some cs ->
+          let cs = CS.remove c cs in
+          if CS.is_empty cs then None else Some cs
+        | None -> None
       ) cs_map
 
-  let subst ex rr nrr bcs =
-    match MX.find rr bcs.cs_map with
-    | ids ->
-      let cs_map, cs_set, fresh =
-        CS.fold (fun cs (cs_map, cs_set, fresh) ->
-            let fresh = CS.remove cs fresh in
-            let cs_set = CS.remove cs cs_set in
-            let cs_map = Constraint.fold_leaves (cs_remove cs) cs cs_map in
-            let cs' = Constraint.subst ex rr nrr cs in
-            if CS.mem cs' cs_set then
-              cs_map, cs_set, fresh
-            else
-              let cs_set = CS.add cs' cs_set in
-              let cs_map = Constraint.fold_leaves (cs_add cs') cs' cs_map in
-              (cs_map, cs_set, CS.add cs' fresh)
-          ) ids (bcs.cs_map, bcs.cs_set, bcs.fresh)
-      in
-      assert (not (MX.mem rr cs_map));
-      { cs_set ; cs_map ; fresh }
-    | exception Not_found -> bcs
-
-  let add bcs c =
-    if CS.mem c bcs.cs_set then
-      bcs
-    else
-      let cs_set = CS.add c bcs.cs_set in
-      let cs_map = Constraint.fold_leaves (cs_add c) c bcs.cs_map in
-      let fresh = CS.add c bcs.fresh in
-      { cs_set ; cs_map ; fresh }
-
-  let fold_r f bcs acc =
-    MX.fold (fun r _ acc -> f r acc) bcs.cs_map acc
-
-  let propagate bcs r dom =
-    match MX.find r bcs.cs_map with
-    | cs -> CS.fold Constraint.propagate cs dom
-    | exception Not_found -> dom
-
-  let propagate_fresh bcs dom =
-    let dom = CS.fold Constraint.propagate bcs.fresh dom in
-    { bcs with fresh = CS.empty }, dom
+  let remove c t =
+    let active = CS.remove c t.active in
+    let args_map =
+      Constraint.fold_args (cs_remove c) c.value t.args_map
+    in
+    let leaves_map = fold_leaves (cs_remove c) c.value t.leaves_map in
+    let pending = CS.remove c t.pending in
+    { active; args_map; leaves_map; pending }
+
+  let subst ~ex rr nrr t =
+    match MX.find rr t.leaves_map with
```

Comment:
Can you explain why is it correct to look up in `t.leaves_map` instead of `t.args_map`?
I guess it's because we perform substitution on leaves?

Reply:
By definition, substitutions apply on leaves (terms and AC applications). It would be incorrect to look in `t.args_map`: for instance, if C is the constraint `r = x & (y ++ z)` (where `++` is BV concatenation) then the `args_map` is `r → C; x → C; (y ++ z) → C` but substitutions can only apply to `r`, `x`, `y` and `z`.

Reply:
I will clarify this because it's not strictly correct: Alt-Ergo does propagate `Subst` rewritings for superterms; for instance, if we substitute `z → z'` then the UF module will generate the substitution `y ++ z → y ++ z'`. However, the Shostak modules only supports substituting leaves, and we ultimately call `X.subst` here.

Now that I think about it more, I believe this was an error in the design of the original `Constraint` module — it should not call `X.subst` but rather simply perform a comparison on the result of congruence closure (i.e. the thing that gets substituted). This is probably also true of the `Domains` module.

Reply:
My connection is terrible, I will try again to comment...
So, as you changed the design of these modules in recent PR, I believe we can merge this one now?